cmd/kubeadm/app/preflight/checks.go

```diff
@@ -0,0 +1,207 @@
+// returned.
+type ServiceCheck struct {
+	service string
+}
+
+func (sc ServiceCheck) Check() (warnings, errors []error) {
+
+	initSystem := getInitSystem()
+	if initSystem == nil {
+		return []error{fmt.Errorf("no supported init system detected, skipping service checks for %s", sc.service)}, nil
+	}
+
+	warnings = []error{}
+
+	if !initSystem.ServiceExists(sc.service) {
+		warnings = append(warnings, fmt.Errorf("%s service does not exist", sc.service))
+		return warnings, nil
+	}
+
+	if !initSystem.ServiceIsEnabled(sc.service) {
+		warnings = append(warnings,
+			fmt.Errorf("%s service is not enabled, please run 'systemctl enable %s.service'",
+				sc.service, sc.service))
+	}
+
+	if !initSystem.ServiceIsActive(sc.service) {
+		errors = append(errors,
+			fmt.Errorf("%s service is not active, please run 'systemctl start %s.service'",
+				sc.service, sc.service))
+	}
+
+	return warnings, nil
+}
+
+// PortOpenCheck ensures the given port is available for use.
+type PortOpenCheck struct {
+	port int
+}
+
+func (poc PortOpenCheck) Check() (warnings, errors []error) {
+	errors = []error{}
+	// TODO: Get IP from KubeadmConfig
+	ln, err := net.Listen("tcp", fmt.Sprintf(":%d", poc.port))
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Port %d is in use", poc.port))
+	}
+	if ln != nil {
+		ln.Close()
+	}
+
+	return nil, errors
+}
+
+// DirAvailableCheck checks if the given directory either does not exist, or
+// is empty.
+type DirAvailableCheck struct {
+	path string
+}
+
+func (dac DirAvailableCheck) Check() (warnings, errors []error) {
+	errors = []error{}
+	// If it doesn't exist we are good:
+	if _, err := os.Stat(dac.path); os.IsNotExist(err) {
+		return nil, nil
+	}
+
+	f, err := os.Open(dac.path)
+	if err != nil {
+		errors = append(errors, fmt.Errorf("Unable to check if %s is empty: %s", dac.path, err))
+		return nil, errors
+	}
+	defer f.Close()
+
+	_, err = f.Readdirnames(1)
+	if err != io.EOF {
+		errors = append(errors, fmt.Errorf("%s is not empty", dac.path))
+	}
+
+	return nil, errors
+}
+
+// InPathChecks checks if the given executable is present in the path.
+type InPathCheck struct {
+	executable string
+	mandatory  bool
+}
+
+func (ipc InPathCheck) Check() (warnings, errors []error) {
+	_, err := exec.LookPath(ipc.executable)
+	if err != nil {
+		if ipc.mandatory {
+			// Return as an error:
+			return nil, []error{fmt.Errorf("%s not found in system path", ipc.executable)}
+		}
+		// Return as a warning:
+		return []error{fmt.Errorf("%s not found in system path", ipc.executable)}, nil
+	}
+	return nil, nil
+}
+
+func RunInitMasterChecks() error {
+	// TODO: Some of these ports should come from kubeadm config eventually:
+	checks := []PreFlightCheck{
+		ServiceCheck{service: "kubelet"},
+		ServiceCheck{service: "docker"},
+		PortOpenCheck{port: 443},
+		PortOpenCheck{port: 2379},
+		PortOpenCheck{port: 8080},
+		PortOpenCheck{port: 10250},
+		PortOpenCheck{port: 10251},
+		PortOpenCheck{port: 10252},
```

Comment:
Firewalling can be done at the host level or with networking equipment we can't control. However, we need to make sure that whatever connects to this ports has proper error-handling in the case these ports are not _open_ for service.